# 錯誤處理機制

- message: 錯誤提示信息
- name: 錯誤名稱（非標準屬性）
- stack: 錯誤的堆棧（非標準屬性）

### SyntaxError 對象

`SyntaxError` 對象是解析代碼時發聲的語法錯誤。

```js
// 变量名错误
var 1a;
// Uncaught SyntaxError: Invalid or unexpected token

// 缺少括号
console.log 'hello');
// Uncaught SyntaxError: Unexpected string
```

上面代碼的錯誤，都是在語法解析階段就可以發現，所以會拋出`SyntaxError`。第一個錯誤提示是 “token 非法”；第二個錯誤提示是“字符串不符合要求”

### ReferenceError 對象

`ReferenceError` 對象是引用一個不存在的變量時發生的錯誤。

```js
// 使用一个不存在的变量
unknownVariable;
// Uncaught ReferenceError: unknownVariable is not defined
```

另一種觸發場景是，講一個值分配給無法分配的對象，比如對函數的運行結果或者 `this` 賦值。

```js
// 等号左侧不是变量
console.log() = 1
// Uncaught ReferenceError: Invalid left-hand side in assignment

// this 对象不能手动赋值
this = 1
// ReferenceError: Invalid left-hand side in assignment
```

上面代碼對函數 `console.log` 的運行結果和 `this` 賦值，結果都引發了 `ReferenceError` 錯誤。

### RangeError 對象

`RangeError`對象是一個值超出有效範圍時發生的錯誤。主要有幾種情況，一是數組長度爲負數，而是 `Number` 對象的方法參數超出範圍，以及函數堆棧超過最大值。

```js
// 數組長度不得爲負數
new Array(-1);
// VM205:1 Uncaught RangeError: Invalid array lengthat                          <anonymous>:1:1
```

### TypeError 對象

`TypeError` 對象是變量或參數不是迂曲類型時發生的錯誤。比如，對字符串、布爾值、數值等原始類型的值使用`new`命令，就會拋出這種錯誤，因爲 `new` 命令的參數應該是一個構造函數。

```js
new 123();
// Uncaught TypeError: 123 is not a constructor at <anonymous>:1:1
```
